feat(internal/client): Implement `CompareAndSetBlockData` on `ClientAdapter`

haikoschol opened the pull request:
## Changes

This PR adds an implementation of `CompareAndSetBlockData()` on `ClientAdapter` using `OffchainStorage`.

## Tests

```sh
go test github.com/ChainSafe/gossamer/internal/client/adapter
```

## Issues

#4468

haikoschol committed:
feat(internal/client): Implement CompareAndSetBlockData on ClientAdapter

haikoschol commented:
@timwu20 @dimartiro Although `OffchainStorage` is accessible from `api.Backend`, it isn't from `Client`. Therefore I've added a method to the `Client` interface in the `adapter` package and to the `Client` implementation in the `client` package. Is that a reasonable approach?

I could also add `api.Backend` to the `Client` interface and forward all of its methods on the `Client`. Or is there an obvious way I've overlooked?

dimartiro commented on internal/client/adapter/client_adapter.go:
```diff
@@ -412,7 +415,20 @@ func (ca *ClientAdapter[H, Hasher, N, E, Header]) HandleRuntimeChanges(newState
 
 func (ca *ClientAdapter[H, Hasher, N, E, Header]) CompareAndSetBlockData(bd *types.BlockData) error {
-	panic("unimplemented")
+	storage := ca.client.OffchainStorage()
```
Maybe the `CompareAndSetBlockData` should be part of `Client` too, so you can use `ClientAdapter` just as a proxy and you won't need to expose the `OffchainStorage` from the client. wdyt?

haikoschol replied:
Sounds good to me. I'll change it.

haikoschol committed:
move CompareAndSetBlockData() to Client

ClientAdapter just proxies the method, alleviating the need
to expose OfflineStorage from the underlying backend.

haikoschol committed:
Merge branch 'development' into haiko/CompareAndSetBlockData

haikoschol committed:
lint

haikoschol commented:
The remaining DeepSource failures are all triggered by unimplemented methods. They will go away once we implement them, so I'm leaving them as is since the check is not required to merge.

haikoschol committed:
Merge branch 'development' into haiko/CompareAndSetBlockData

haikoschol committed:
Merge branch 'development' into haiko/CompareAndSetBlockData

haikoschol committed:
Merge branch 'development' into haiko/CompareAndSetBlockData

haikoschol committed:
fixed botched conflict resolution

haikoschol committed:
Merge branch 'development' into haiko/CompareAndSetBlockData

dimartiro reviewed (approved):
LGTM
